uaclient/cli.py

```diff
@@ -1114,22 +1136,70 @@ def action_auto_attach(args, *, cfg):
 @assert_root
 @assert_lock_file("ua attach")
 def action_attach(args, *, cfg):
-    if not args.token:
+    if not args.token and not args.attach_config:
         raise exceptions.UserFacingError(
             ua_status.MESSAGE_ATTACH_REQUIRES_TOKEN
         )
-    try:
-        actions.attach_with_token(
-            cfg, token=args.token, allow_enable=args.auto_enable
+    if args.token and args.attach_config:
+        raise exceptions.UserFacingError(
+            ua_status.MESSAGE_ATTACH_TOKEN_ARG_XOR_CONFIG
         )
+
+    if args.token:
+        token = args.token
+        enable_services_override = None
+    else:
+        try:
+            attach_config = AttachActionsConfigFile.from_dict(
+                yaml.safe_load(args.attach_config)
+            )
+        except IncorrectTypeError as e:
+            raise exceptions.UserFacingError(
+                textwrap.fill(
+                    "Error while reading {}: {}".format(
+                        args.attach_config.name, e.msg
+                    ),
+                    width=PRINT_WRAP_WIDTH,
+                )
+            )
+        token = attach_config.token
+        enable_services_override = attach_config.enable_services
+
+    allow_enable = args.auto_enable and enable_services_override is None
+
+    try:
+        actions.attach_with_token(cfg, token=token, allow_enable=allow_enable)
     except util.UrlError:
         event.info(ua_status.MESSAGE_ATTACH_FAILURE)
         return 1
     except exceptions.UserFacingError:
         return 1
     else:
+        ret = 0
+        if enable_services_override is not None and args.auto_enable:
+            found, not_found = get_valid_entitlement_names(
+                enable_services_override
+            )
+            for name in found:
+                ent_ret, reason = _enable(
+                    cfg, name, assume_yes=True, allow_beta=True
+                )
+                if not ent_ret:
+                    ret = 1
+                    if (
+                        reason is not None
+                        and isinstance(reason, ua_status.CanEnableFailure)
+                        and reason.message is not None
+                    ):
+                        event.info(reason.message)
+            if not_found:
+                msg = _create_enable_entitlements_not_found_message(
+                    not_found, allow_beta=True
+                )
+                logging.warning(msg)
```

Comment:
This will print to stdout as well, right ? If yes, I think we should consider using the event logic here

Reply:
Yes, and I thought of that but I think it will make more sense to leave it as is in this PR. Then, in the PR that adds `--format=json` support to `attach`, we can change this to use the event_logger appropriately

Reply:
Fair enough